src/style/style.js

```diff
@@ -568,7 +568,11 @@ class Style extends Evented {
 
         layer.setEventedParent(this, {layer: {id: id}});
 
-        const index = before ? this._order.indexOf(before) : this._order.length;
+
+        const beforeIndex = before ? this._order.indexOf(before) : -1;
+        const index = before && beforeIndex !== -1 ? beforeIndex : this._order.length;
+        if (before && beforeIndex === -1) util.warnOnce(`Layer with id "${before}" does not exist on this map. Adding layer to the top of the stack.`);
```

Comment:
Let's report this via `this.fire('error', ...)` for consistency with non-existent layer IDs in `moveLayer`, `removeLayer`, etc.